File: solutions/python/bank-account/1/bank_account.py
class BankAccount:
    def __init__(self):
        self._balance = 0
        self._is_open = False

    def _ensure_open(self):
        if not self._is_open:
            raise ValueError("account not open")

    def get_balance(self):
        self._ensure_open()
        return self._balance

    def open(self):
        if self._is_open:
            raise ValueError("account already open")

        self._balance = 0
        self._is_open = True

    def deposit(self, amount):
        self._ensure_open()
        if amount <= 0:
            raise ValueError("amount must be greater than 0")
        self._balance += amount

    def withdraw(self, amount):
        self._ensure_open()
        if amount <= 0:
            raise ValueError("amount must be greater than 0")
        if amount > self._balance:
            raise ValueError("amount must be less than balance")
        self._balance -= amount

    def close(self):
        self._ensure_open()
        self._is_open = False
        self._balance = 0

File: solutions/cpp/bank-account/1/bank_account.cpp
#include "bank_account.h"

#include <stdexcept>

namespace Bankaccount {
    Bankaccount::Bankaccount() : balance_(0) {
    }

    void Bankaccount::open() {
        std::lock_guard<std::mutex> lock(mtx);
        if (!name.empty()) {
            throw std::runtime_error("Account already open");
        }
        name.append(std::to_string(rand() % 10001));
    }

    int Bankaccount::balance() const {
        std::lock_guard<std::mutex> lock(mtx);
        if (name.empty()) {
            throw std::runtime_error("Account not open");
        }
        return balance_;
    }

    void Bankaccount::deposit(const int deposit) {
        std::lock_guard<std::mutex> lock(mtx);
        if ( deposit < 1 || name.empty()) {
            throw std::runtime_error("Deposit must be greater than 0.");
        }
        balance_ += deposit;
    }

    void Bankaccount::withdraw(const int debt) {
        std::lock_guard<std::mutex> lock(mtx);
        if (debt < 1 || name.empty() || debt > balance_) {
            throw std::runtime_error("error");
        }
        balance_ -= debt;
    }

    void Bankaccount::close() {
        std::lock_guard<std::mutex> lock(mtx);
        if (name.empty()) {
            throw std::runtime_error("Account not open");
        }
        name="";
        balance_ = 0;
    }
}
